### PR TITLE
format doctests in docstrings

### LINKED ISSUE #7146: Using ruff to format code examples in docstrings
At @pola-rs , we currently use `ruff` in combination with `black` for our Python formatting needs. We also use [`blackdoc`](https://github.com/keewis/blackdoc) for formatting our docstring code examples. Out of these three, `blackdoc` is by far the slowest.

With `ruff` gaining many of the capabilities of the `black` formatter, it would be great if we could also replace `blackdoc` soon and only use `ruff`.

### PR DESCRIPTION
**Pull request by BurntSushi**
## Summary

This PR adds opt-in support for formatting doctests in docstrings. This reflects initial support and it is intended to add support for Markdown and reStructuredText Python code blocks in the future. But I believe this PR lays the groundwork, and future additions for Markdown and reST should be less costly to add.

It's strongly recommended to review this PR commit-by-commit. The last few commits in particular implement the bulk of the work here and represent the denser portions.

Some things worth mentioning:

* The formatter is itself not perfect, and it is possible for it to produce invalid Python code. Because of this, reformatted code snippets are checked for Python validity. If they aren't valid, then we (unfortunately silently) bail on formatting that code snippet.
* There are a couple places where it would be nice to at least warn the user that doctest formatting failed, but it wasn't clear to me what the best way to do that is.
* I haven't yet run this in anger on a real world code base. I think that should happen before merging.

Closes #7146 

## Test Plan

* [x] Pass the local test suite.
* [x] Scrutinize ecosystem changes.
* [x] Run this formatter on extant code and scrutinize the results. (e.g., CPython, numpy.)

### CODE REVIEW

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 1]
```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
+        match self.code_example.add(line) {
+            CodeExampleAddAction::Print { original } => self.print_one(&original)?,
+            CodeExampleAddAction::Kept => {}
+            CodeExampleAddAction::Format {
+                kind,
+                code,
+                original,
+            } => {
+                let Some(formatted_lines) = self.format(&code)? else {
+                    // If formatting failed in a way that should not be loud,
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/expression/string.rs:1214]: Nit: "should not be allowed"

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    pass
+
+
+# Tests that long lines get wrapped... appropriately.
+#
+# The docstring code formatter uses the same line width settings as for
+# formatting other code. This means that a line in the docstring can
+# actually extend past the configured line limit.
+#
+# It's not quite clear whether this is desirable or not. We could in
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: Yeah I'm not sure. I suspect the behavior you have here _is_ desirable? (Any idea what other tools do?)

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+     >>> def cool_stuff( x ):
+     ...     print( x )
+    ...     print( f"hi {x}" );
+    hi 2
+    """
+    pass
+
+
+# Checks that a doctest with improper triple single
+# quoted string gets skipped.
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:275]: This is getting skipped because the doctest fails to parse, right? As in, it's invalid Python?

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 4]
```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
+    Ok(formatted.print()?)
+}
+
+/// If the last line of the docstring is `content" """` or `content\ """`, we need a chaperone space
+/// that avoids `content""""` and `content\"""`. This does only applies to un-escaped backslashes,
+/// so `content\\ """` doesn't need a space while `content\\\ """` does.
+fn needs_chaperone_space(normalized: &NormalizedString, trim_end: &str) -> bool {
+    trim_end.ends_with(normalized.quotes.style.as_char())
+        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
 }
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/expression/string.rs:1678]: Slightly superficial comment, but what's your reaction to splitting the docstring-specific code out into its own file?

File: crates/ruff_workspace/src/settings.rs [hunk 5]
```diff
@@ -124,6 +126,8 @@ pub struct FormatterSettings {
     pub magic_trailing_comma: MagicTrailingComma,
 
     pub line_ending: LineEnding,
+
+    pub docstring_code: DocstringCode,
```
Comment (charliermarsh) [inline crates/ruff_workspace/src/settings.rs:130]: I feel like we're going to end up bikeshedding on this name... a few ideas for consideration:

- `include_docstrings=true` (I like this but it doesn't convey that we're describing _code_ in docstrings)
- `docstrings=true`
- `include_embedded_code=true`
- `include_docstring_code=true`
- `include_docstring_snippets=true`

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 6]
```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
+    /// to the formatter itself returns a `FormatError`. In all other cases
+    /// (for example, if the code snippet is invalid Python or even if the
+    /// resulting reformatted code snippet is invalid Python), then `Ok(None)`
+    /// is returned. In this case, callers should assume that a reformatted
+    /// code snippet is unavailable and bail out of trying to format it.
+    ///
+    /// Currently, when the above cases happen and `Ok(None)` is returned, the
+    /// routine is silent about it. So from the user's perspective, this will
+    /// fail silently. Ideally, this would at least emit a warning message,
+    /// but at time of writing, it wasn't clear to me how to best do that.
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/expression/string.rs:1322]: Hmm yeah. I guess the warning utilities only exist in the linter. We could move those macros to (e.g.) `ruff_macro`, or something, but I'm not hugely concerned about this.

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 7]
```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
+
+        let tab_or_non_ascii_space = trim_end
+            .chars()
+            .take_while(|c| c.is_whitespace())
+            .any(|c| c != ' ');
+
+        if tab_or_non_ascii_space {
+            // We strip the indentation that is shared with the docstring
+            // statement, unless a line was indented less than the docstring
+            // statement, in which we strip only this much indentation to
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/expression/string.rs:1274]: Nit: "in which case" (I know this was just moved, but I re-read it so...)

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 8]
```diff
@@ -1080,74 +1144,537 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
+        let last_line_is_last = code
+            .last()
+            .expect("code blob must be non-empty")
+            .original
+            .is_last;
+        let codeblob = code
+            .iter()
+            .map(|line| &*line.code)
+            .collect::<Vec<&str>>()
+            .join("\n");
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/expression/string.rs:1343]: Hmm, we may need to preserve line endings here, right? What if the code uses non-LF line endings, or the formatter has a line ending set?

**Review by charliermarsh — COMMENTED** [review]
Overall, this looks great, and I appreciate the way you broke it down by commit. I ended up reading the commits one-by-one and found the flow very clear.

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py (hunk 2 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: So it looks like `blacken-docs` behaves like us. But [`blackdoc` passes the current indentation width to `black`](https://github.com/keewis/blackdoc/blob/822385ba8955e44397288caa07a2d2e3dc933b1e/blackdoc/blacken.py#L72) when reformatting a code snippet. From playing with it, it does indeed look like that the max line width is preserved uniformly.

I will also say that as someone who cares a lot about line width, I would want the formatter to wrap lines according to my setting in a global way. i.e., The code snippets shouldn't be treated as starting from the first column.

I'll take a quick look and see if I can make this work _robustly_. (Emphasis on "robust." This was why I didn't explore it too much initially.)

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py (hunk 3 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:275]: Yup! I added a clarifying note to the comment.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 4 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1678]: I thought about this while coding and I was torn. I think I ultimately didn't do it because I thought that we might eventually want to refactor it further (possibly) so that it can be used in other contexts aside from formatting Python code. So I didn't want to spend too much time on making it a "proper" abstraction. With that said, I do believe it lifts pretty cleanly out of this module. I'll try it out and see how it feels.

File: crates/ruff_workspace/src/settings.rs (hunk 5 above)
Comment (BurntSushi) [reply crates/ruff_workspace/src/settings.rs:130]: Yeah that was why I just picked the first thing that came to mind.

Of your list, I like `include_docstring_code` the best. But I think I do actually like `docstring_code` the best. It kind of feels like the word "include" is a little redundant? Looking at the names of other options, it doesn't look like there is any strong precedent here. Although using "include" here could start a precedent. (I don't know whether we plan on adding a ton of options or not though.)

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 4 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/expression/string.rs:1678]: It would also be fine as a separate change / PR after-the-fact if easier and less distracting.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 6 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1322]: All righty. I'll keep it out of scope for now. cc @zanieb since I think you might care about this and might have other ideas  or opinions.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 8 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1343]: I don't believe so, no. I could use `\r\n` here and things would still turn out okay. I did it this way because immediately after this, we reformat the code snippet with the user's settings, which should include their line ending preferences.

With that said... I tried to write a test for this and failed. I _suspect_ something is mangling or normalizing line endings in our test infrastructure. If I run the formatter directly, then line endings are preserved:

```
$ rg '\x0D' /tmp/test.py
$ cat /tmp/ruff.toml
[format]
docstring-code = true
$ cargo run -qp ruff_cli format --config /tmp/ruff.toml - < /tmp/test.py | rg '\x0D'
$ cat /tmp/ruff.toml
[format]
docstring-code = true
line-ending = 'cr-lf'
$ cargo run -qp ruff_cli format --config /tmp/ruff.toml - < /tmp/test.py | rg '\x0D'
def doctest_line_ending():
    """
    Do cool stuff.
    >>> def foo(x):
    ...     print(x)
    ...
    ...     print(x)
    """
    pass
```

I'll take a look and see if I can figure out what's swallowing line endings. When I add a test with `line-ending = "CarriageReturnLineFeed"`, the resulting snapshot does not contain CRLF line endings, but LF only.

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/expression/string.rs:1343]: Take a look at the `.gitattributes` in the repo root. I believe we normalize to LF except in explicit cases.

File: crates/ruff_workspace/src/settings.rs (hunk 5 above)
Comment (zanieb) [reply crates/ruff_workspace/src/settings.rs:130]: I was okay with this on the first read, but on a second, I think `docstring_code = true | false` doesn't have enough meaning alone. Perhaps `format_docstring_code` would make sense? I also don't mind `include_docstring_code` but it's not necessarily clear what inclusion means.

Comment (BurntSushi) [reply crates/ruff_workspace/src/settings.rs:130]: Yeah I hadn't gone with `format_docstring_code` since the option exists in the context of a formatter. But I think I do like that better than `include_docstring_code`.

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py (hunk 2 above)
Comment (zanieb) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: A separate setting _might_ make sense e.g. `docstring_code_line_width`. I can see arguments for both sides. For me, I care more about a reasonable width in my reference documentation generated from my docstrings and would be okay with starting on the first column.

On a slightly separate note, `docstring_code_...` settings suggest that `docstring_code_enabled` may be a better name for toggling this behavior?

Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: Yeah, that's a good point.

Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: OK, here is what I suggest:

* We keep the current behavior which treats the code snippet as starting at column 1. This I think aligns with what @zanieb said about prioritizing the display of the code snippet.
* We create an issue tracking a new feature for overriding this behavior to do something different.

Probably `docstring_code_line_width` would be a good thing to add, but I think there's also a different mode of "respect the top-level line width setting." I don't think you can get that behavior by specifying a distinct concrete docstring code line width, since the formatter needs to take indentation into account.

So basically, there's some implementation work here to change the behavior and also some decisions that need to be made about how we present this behavior.

If this sounds good to y'all I can create a new issue capturing this.

File: crates/ruff_workspace/src/settings.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/ruff_workspace/src/settings.rs:130]: I still feel like none of these are quite right (the docstring itself _is_ code)... It's a really hard thing to convey succinctly. I wish we could do `format_doctests` but this needs to generalize to embedded Markdown and rST. Is `format_docstring_snippets` clearer or worse? (At the end of the day, I can live with `format_docstring_code`.)

**Comment by BurntSushi** [comment]
I ran the docstring code formatter on CPython (after formatting the code _without_ docstring code formatting enabled) and put the results in one commit: https://github.com/BurntSushi/cpython/commit/19d1c85db9c862172f0cffd0872a942aef26d934

I did a quick pass through them and everything looks like it passes the eyeball test. My next step is to make sure the doctests still run and pass.

**Comment by BurntSushi** [comment]
> My next step is to make sure the doctests still run and pass.

I checked out the `main` branch of CPython and built it:

```
$ export CFLAGS="${CFLAGS/-O2/-O3} -ffat-lto-objects"
$ ./configure --prefix=/usr \
              --enable-shared \
              --with-computed-gotos \
              --enable-optimizations \
              --with-lto \
              --enable-ipv6 \
              --with-system-expat \
              --with-dbmliborder=gdbm:ndbm \
              --with-system-libmpdec \
              --enable-loadable-sqlite-extensions \
              --without-ensurepip \
              --with-tzpath=/usr/share/zoneinfo
$ LC_CTYPE=en_US.UTF-8 make -j8 EXTRA_CFLAGS="$CFLAGS"
```

Then confirmed that I could run some doctests:

```
$ ./python -m doctest Lib/ipaddress.py -v
```

And some of that failed, but the [`summarize_address_range`](https://github.com/BurntSushi/cpython/commit/19d1c85db9c862172f0cffd0872a942aef26d934#diff-f67a8b5be6bdc48383d2ab907325a4c75ef998fa7cd0bba382ec2efa7e4f00b5R204-R205jj) doctest passed.

Then I checked out my branch with the reformatted code snippets and re-ran the above. The results remain unchanged.

So at least in that one specific case that looked potentially odd, reformatting didn't break the doctest. (Specifically, the doctest directive in this example is still picked up.)

I haven't been able to figure out how to run all of the doctests yet though.

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 9]
```diff
@@ -1160,13 +1299,99 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+            Err(FormatModuleError::FormatError(err)) => return Err(err),
+            Err(
+                FormatModuleError::LexError(_)
+                | FormatModuleError::ParseError(_)
+                | FormatModuleError::PrintError(_),
+            ) => {
+                return Ok(None);
+            }
+        };
+        // This is a little hokey, but we want to determine whether the
```
Comment (konstin) [inline crates/ruff_python_formatter/src/expression/string.rs:1355]: We should create an issue for removing this, parsing twice every time seems wasteful

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py (hunk 2 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: So one thing I uncovered today is that by allowing docstring lines to exceed the globally configured line width, you wind up with linter errors on long lines. For example, after running the docstring code snippet formatter on polars: https://github.com/BurntSushi/polars/actions/runs/7006619426/job/19058868021?pr=1

**Comment by BurntSushi** [comment]
I've split the commit that adds a user facing config option out into #8854 so that we should be able to merge this PR without any user facing changes.

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py (hunk 2 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/docstring_code_examples.py:199]: Issue created: https://github.com/astral-sh/ruff/issues/8855

File: crates/ruff_workspace/src/settings.rs (hunk 5 above)
Comment (BurntSushi) [reply crates/ruff_workspace/src/settings.rs]: PR for the user facing change created (and removed from this PR): https://github.com/astral-sh/ruff/pull/8854

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 6 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1196]: Issue created: https://github.com/astral-sh/ruff/issues/8856

File: crates/ruff_workspace/src/options.rs [hunk 10]
```diff
@@ -2795,6 +2795,49 @@ pub struct FormatOptions {
...
+    ///     Something about `f`. And an example:
+    ///
+    ///     >>> f(x)
+    ///     """
+    ///     pass
+    /// ```
+    ///
+    /// If a doctest contains invalid Python code or if the formatter would
+    /// otherwise write invalid Python code, then the doctest is ignored by
+    /// the formatter and kept as-is.
```
[the remark this answers is not in the document]
Comment (BurntSushi) [reply crates/ruff_workspace/src/options.rs]: Carried this over to #8854: https://github.com/astral-sh/ruff/pull/8854#issuecomment-1828141780

**Comment by BurntSushi** [comment]
All righty, I've filed issues for problems that probably aren't blocking this PR being merged:

* https://github.com/astral-sh/ruff/issues/8855
* https://github.com/astral-sh/ruff/issues/8856
* https://github.com/astral-sh/ruff/issues/8857

I'm going to go ahead and merge this PR. I'd be happy to address any other feedback folks might have!

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 11]
```diff
@@ -1080,74 +1075,105 @@ fn format_docstring(normalized: &NormalizedString, f: &mut PyFormatter) -> Forma
...
-    trim_end.ends_with(normalized.quotes.style.as_char())
-        || trim_end.chars().rev().take_while(|c| *c == '\\').count() % 2 == 1
-}
-
-/// Format a docstring line that is not the first line
-fn format_docstring_line(
-    line: &str,
-    is_last: bool,
+/// An abstraction for printing each line of a docstring.
+struct DocstringLinePrinter<'ast, 'buf, 'fmt> {
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs]: I think the way it is implemented is fine, considering that `print` takes `lines` as an argument. 

The alternative (that @konstin) suggested is to instead pass `lines` to the `DocstringLinePrinter` constructor and implement `Format` for `DocstringLinePrinter`. I would then probably rename the struct to `DocstringLines`, because it is now a struct that knows how to format itself. 

However, I believe the problem with that is that `Format::format` takes `&self` and not `&mut self`, making it impossible for the printer to update its internal fields (without using internal mutability)

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 12]
```diff
@@ -1043,6 +1043,80 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+
+/// Represents a single line in a docstring.
+///
+/// This type is used to both represent the original lines in a docstring
+/// (the line will be borrowed) and also the newly formatted lines from code
+/// snippets (the line will be owned).
+#[derive(Clone, Debug)]
+struct DocstringLine<'src> {
+    /// The actual text of the line, not including the line terminator.
+    line: Cow<'src, str>,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1278]: It took me a while to understand when the line is owned or borrowed although it is well documented. It may help to repeat the struct documentation in a short form:

The actual text (Borrowed: original line in the source, Owned: formatted line) of the line

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 13]
```diff
@@ -1043,6 +1043,80 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum CodeExampleKind {
+    /// Code found in Python "doctests."
+    ///
+    /// Documentation describing doctests and how they're recognized can be
+    /// found as part of the Python standard library:
+    /// https://docs.python.org/3/library/doctest.html.
+    ///
+    /// (You'll likely need to read the regex matching used internally by the
+    /// doctest module to determine more precisely how it works.)
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1383]: Could we add a link to where this regex can be found?

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 14]
```diff
@@ -1043,6 +1043,80 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+}
+
+/// State corresponding to a single doctest code example found in a docstring.
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct CodeExampleDoctest {
+    /// The indent observed in the first doctest line.
+    ///
+    /// More precisely, this corresponds to the whitespace observed before
+    /// the starting `>>> ` (the "PS1 prompt").
+    indent: String,
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1394]: Future perf improvement: Represent as an enum with `Tabs(count)`, `Spaces(count)`, and `Mixed(String)` to avoid allocating for each formatted line. 

Or could the indent be a ` &'a str` because it is the indent observed from the source text?

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 15]
```diff
@@ -1117,6 +1194,97 @@ struct CodeExampleLine<'src> {
...
+    /// callers should pass it through to the formatter as-is.
+    Format {
+        /// The kind of code example that was found.
+        kind: CodeExampleKind,
+        /// The Python code that should be formatted, indented and printed.
+        ///
+        /// This is guaranteed to be non-empty.
+        code: Vec<CodeExampleLine<'src>>,
+        /// When set, the line is considered not part of any code example
+        /// and should be formatted as if the `Ignore` action were returned.
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1453]: Should `Ignore` be the `Print` action? Maybe use [`Ignore`](CodeExampleAddAction::Print) so that rustfmt can point out the invalid refernece.

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 16]
```diff
@@ -1071,6 +1071,83 @@ impl<'src> DocstringLine<'src> {
...
+    /// # Panics
+    ///
+    /// This panics when the existing code-example is any non-None value. That
+    /// is, this routine assumes that there is no ongoing code example being
+    /// collected and looks for the beginning of another code example.
+    fn add_start(&mut self, original: DocstringLine<'src>) -> Option<DocstringLine<'src>> {
+        assert_eq!(None, self.kind, "expected no existing code example");
+        if let Some((indent, code)) = doctest_find_ps1_prompt(&original.line) {
+            let indent = indent.to_string();
+            let code = code.to_string();
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1364]: What's the reason for converting these to `String` rather than storing borrowed `&str`?

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 17]
```diff
@@ -1117,6 +1194,97 @@ struct CodeExampleLine<'src> {
...
+/// `None` will be returned).
+fn doctest_find_ps2_prompt<'src>(ps1_indent: &str, line: &'src str) -> Option<&'src str> {
+    let (ps2_indent, ps2_after) = line.split_once("...")?;
+    // PS2 prompts must have the same indentation as their
+    // corresponding PS1 prompt.[1] While the 'doctest' Python
+    // module will error in this case, we just treat this line as a
+    // non-doctest line.
+    //
+    // [1]: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L733
+    if ps1_indent != ps2_indent {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1497]: I addmit I haven't read the code in full, but would it be sufficient to check if the indents have the same column/character length as done in the Lexer or is it necessary that the strings match exactly?

See [`Indentation`](https://github.com/astral-sh/ruff/blob/8ce138760a182b4765fa6f57b865fe6e9507f41e/crates/ruff_python_parser/src/lexer/indentation.rs#L35-L80) in our lexer.

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 18]
```diff
@@ -1037,13 +1173,99 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+        let last_line_is_last = code
+            .last()
+            .expect("code blob must be non-empty")
+            .original
+            .is_last;
+        let codeblob = code
+            .iter()
+            .map(|line| &*line.code)
+            .collect::<Vec<&str>>()
+            .join("\n");
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1217]: How about line endings inside of multiline strings? Ideally, we would preserve these. Although we may deliberately decide not to and encourage users to instead use explicit line endings. But changing the line endings inside of strings is theoretically a semantic change.

File: crates/ruff_python_formatter/src/expression/string.rs [hunk 19]
```diff
@@ -1037,13 +1173,99 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
...
+        // So attempt to parse it as Python code, but ensure it is wrapped
+        // within a docstring using the same quotes as the docstring we're in
+        // right now.
+        //
+        // This is an unfortunate stop-gap to attempt to prevent us from
+        // writing invalid Python due to some oddity of the code snippet within
+        // a docstring. As we fix corner cases over time, we can perhaps
+        // remove this check. See the `doctest_invalid_skipped` tests in
+        // `docstring_code_examples.py` for when this check is relevant.
+        let wrapped = match self.quote_style {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/string.rs:1240]: Could we short-circuit if the source string doesn't contain any triple quotes (or escaped triple quotes)?

**Review by MichaReiser — COMMENTED** [review]
Thanks for writing the excellent documentation and organizing your commits so thoughtfully. It helped a lot when reviewing the PR (also to find savepoints when being interrupted)

Really impressive work. Well done. I've a few smaller comments. The most important one from a correctness point of view is the handling of newlines inside of multiline strings. 


I think I'll learn many new English words from your commit messages 😆. So far: 

* rejigger
* grok

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs]: Aye. I also expect this abstraction to get mangled and refactored at some point in the future. Namely, I expect that it will become useful to have "look for Python code snippets" logic outside of the context of the Python formatter. That is to say, I don't expect this is its final form. :-)

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 12 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1278]: Love it. Good catch.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 13 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1383]: There are some links further down in the actual doctest parsing code, but I've added another one here.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 14 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1394]: I believe the indent could indeed be a `&'src str` here, but I went with the simpler approach because the extra allocation here seems marginal to me. We're already allocating other stuff, invoking the formatter and so on.

The reason why I chose to thread through a lifetime parameter for `DocstringLine` is that it gets created for every line in every docstring, regardless of whether there is a code snippet. So I wanted to be a little more cautious there. But by the time we create an `indent` string, we've already (very likely) committed ourselves to a bunch of extra costs too.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 15 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1453]: Yup, good catch. I had originally named it `Ignore` but thought better of it. Forgot to update the docs referring to it.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 16 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1364]: In addition to [my other answer](https://github.com/astral-sh/ruff/pull/8811#discussion_r1408057410), at least with respect to `code`, I think when I was writing the types it wasn't obvious to me that the code portion of a line would necessarily be a proper substring of it. But I can't think of any contrary case.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 17 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1497]: The Python `doctest` module actually requires that the indent be made purely of ASCII space characters. Then it somewhat circuitously takes the length of the indentation in characters, reconstitutes the indentation as a string and then checks that all PS2 prompt lines have the same indentation prefix: https://github.com/python/cpython/blob/0ff6368519ed7542ad8b443de01108690102420a/Lib/doctest.py#L727-L733

Technically we are a little more relaxed here because we allow any kind of whitespace in the indentation. But I preserved the "indentation should be byte-for-byte equivalent" check.

So bottom line is that it's hard for me to say whether this is necessary or not. Probably not. But I do think it is pretty simple? Are you concerned about the costs of carrying the indentation string around? I think for code snippet formatting it is probably a lot less of a concern than when parsing arbitrary Python. The size scales are different.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 18 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1217]: I think Charlie had a similar comment above. The thinking here is that while I'm using `\n` to assemble a code snippet, that snippet is then fed into the formatter and the line endings used correspond to the user's setting. There is a test for example that this writes out CRLF line endings in code snippets when the user has configured CRLF line endings.

Did I understand you concern correctly here? I feel like I might be missing it.

Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1217]: Also, there are places in the docstring handling (that existed before my changes) that seem to assume line feed terminators? For example:

https://github.com/astral-sh/ruff/blob/2ade84a29de99eafbac23ef4c6dc9cdf1101e014/crates/ruff_python_formatter/src/expression/string/docstring.rs#L237-L238

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 19 above)
Comment (BurntSushi) [reply crates/ruff_python_formatter/src/expression/string.rs:1240]: Yeah I think you mentioned this in the tracking issue for this. I _think_ you're correct. And if so, yes, I believe we could short circuit. I'm not sure though if we want to keep this check as-is here for now as a conservative posture.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 18 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs:1217]: Yeah, I think the problem is not new due to your changes. It only became evident that this might now be a problem when thinking about code snippets in docstrings (it shouldn't be a problem for non-code text). 

```python
a = """A multiline string
with windows line endings
"""

b = c
```
Screenshot 2023-11-29 at 07 23 49

My concern is (was) that changing the newlines **inside** of the multiline string could be a semantic change. I tried to read through the Python documentation and only found:

> In triple-quoted literals, unescaped newlines and quotes are allowed (and are retained), except that three unescaped quotes in a row terminate the literal. (A “quote” is the character used to open the literal, i.e. either ' or ".) [source](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)

What I understand from the spec is that newlines are retained. Although running the above in a `print` on my mac normalizes the newlines to `\n`. 

I then tested what the about for

```python
a = b"""A multiline string
with windows line endings
"""

from binascii import hexlify

print(hexlify(a))
```

and it seems python normalizes the newlines even for binary strings to `\n`. So the retain only means that the newlines are preserved but not in the form they're present in the source code.

So I guess this is not a problem after all (and ruff and Black both normalize newlines inside multiline strings)

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 17 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs:1497]: I was mainly trying to understand the semantics to ensure we are as strict as Python when it comes to handling indentation. Thanks for the explanation.

File: crates/ruff_python_formatter/src/expression/string.rs (hunk 19 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/string.rs:1240]: Understood. My only concern is performance because our parser isn't particularly fast today (I believe it's about 30% or more of the overall formatting time). An alternative could be to only lex the code and see if there are any lexer errors. But I don't know if that's sufficient to detect the kind of errors that could be introduced.

**Comment by JacobCoffee** [comment]
Can #3792, #8237 be closed then?

**Comment by BurntSushi** [comment]
@JacobCoffee Yeah I think those issues are about applying `ruff` to Python code in contexts other than Python source files. Like `.md` or `.rst` files. And I think it also applies to `ruff` generally and not just `ruff format`.
